pgdog/src/frontend/router/parser/cache.rs

```diff
@@ -55,52 +66,45 @@ impl Cache {
     /// parse the same query. That's better imo than locking the data structure
     /// while we parse the query.
     pub fn parse(&mut self, query: &str) -> Result<Arc<ParseResult>> {
-        {
-            let mut guard = self.inner.lock();
-            let ast = guard.queries.get_mut(query).map(|entry| {
-                entry.hits += 1;
-                entry.ast.clone()
-            });
-            if let Some(ast) = ast {
-                guard.stats.hits += 1;
-                return Ok(ast);
-            }
+        if let Some(mut entry) = self.inner.queries.get_mut(query) {
+            entry.hits += 1;
+            self.inner.stats.hits.fetch_add(1, Relaxed);
+            return Ok(entry.ast.clone());
         }
 
         // Parse query without holding lock.
         let entry = CachedAst::new(parse(query)?);
         let ast = entry.ast.clone();
 
-        let mut guard = self.inner.lock();
-        guard.queries.insert(query.to_owned(), entry);
-        guard.stats.misses += 1;
+        self.inner.queries.insert(query.to_owned(), entry);
+        self.inner.stats.misses.fetch_add(1, Relaxed);
 
         Ok(ast)
     }
 
     /// Get global cache instance.
     pub fn get() -> Self {
-        CACHE.clone()
+        CACHE.read().clone()
     }
 
     /// Get cache stats.
     pub fn stats() -> Stats {
-        Self::get().inner.lock().stats
+        Self::get().inner.stats.clone()
     }
 
     /// Get a copy of all queries stored in the cache.
-    pub fn queries() -> HashMap<String, CachedAst> {
-        Self::get().inner.lock().queries.clone()
+    pub fn queries() -> Arc<DashMap<String, CachedAst>> {
+        Self::get().inner.queries.clone()
     }
 
     /// Reset cache.
     pub fn reset() {
-        let cache = Self::get();
-        let mut guard = cache.inner.lock();
-        guard.queries.clear();
-        guard.queries.shrink_to_fit();
-        guard.stats.hits = 0;
-        guard.stats.misses = 0;
+        // This is the only place we acquire the write lock, in order to synchronize clearing the cache
```

Comment:
Why do we need an RwLock at all? Both `stats` and `queries` are synchronized, we can just clear the DashMap and reset the counters to zero. DashMap has its own RwLock, so we're double-locking for each access to the cache.

Reply:
probably just out of an abundance of caution, making this operation atomic seemed nice but I guess its just fine if the stats are a few cycles behind the map.